pkg/gangway/gangway.go

```diff
@@ -170,87 +173,97 @@ func TranslateProwJobType(prowJobType prowcrd.ProwJobType) JobExecutionType {
 	if err != nil {
 		logrus.WithError(err).Debug("could not find client in allowlist")
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+		return &emptypb.Empty{}, status.Error(codes.InvalidArgument, err.Error())
 	}
 
-	// TODO:
-	// All ProwJob need to be listed, because FieldSelectors are not supported by CRDs yet.
-	// Once FieldSelectors are supported (Kubernetes 1.30 maybe), we can filter the ProwJob list by the desired fields.
-	// Issue link: https://github.com/kubernetes/kubernetes/issues/53459
-	pjList, err := gw.ProwJobClient.List(ctx, metav1.ListOptions{})
-	if err != nil {
-		logrus.WithError(err).Errorf("failed to list ProwJobs")
-		return nil, err
-	}
-
-	var errs []error
-	jobExecutions := []*JobExecution{}
-	for _, pj := range pjList.Items {
-		if !isMatchingCondition(pj, request) {
-			continue
+	go func() {
+		labelMap, err := metav1.LabelSelectorAsMap(getRequestLabelSelector(request))
+		if err != nil {
+			logrus.WithError(err).Debug("could not convert label selector to map")
+		}
+		timeoutSeconds := int64(LIST_TIMEOUT) // increasing the timeout for large clusters
+		options := metav1.ListOptions{
+			LabelSelector:  labels.SelectorFromSet(labelMap).String(),
+			TimeoutSeconds: &timeoutSeconds,
+		}
+		// TODO(Prucek):
+		// All ProwJob need to be listed, because FieldSelectors are not supported by CRDs yet.
+		// Once FieldSelectors are supported (Kubernetes 1.30 maybe), we can filter the ProwJob list by the desired fields.
+		// Issue link: https://github.com/kubernetes/kubernetes/issues/53459
+
+		// creating a context that does not get cancelled and finish the task in the background
+		context, cancel := context.WithTimeout(context.WithoutCancel(ctx), CONTEXT_TIMEOUT)
+		// For now we only use label selector, for fields that are also labels.
+		pjList, err := gw.ProwJobClient.List(context, options)
+		if err != nil {
+			logrus.WithError(err).Errorf("failed to list ProwJobs")
 		}
-		if allowedApiClient != nil {
-			authorized := ClientAuthorized(allowedApiClient, pj)
-			if !authorized {
-				logrus.Error("client is not authorized to modify the given job")
-				errs = append(errs, status.Errorf(codes.PermissionDenied, "client is not authorized to modify the given job: %s", pj.Name))
+		for _, pj := range pjList.Items {
+			if !isMatchingCondition(pj, request) {
 				continue
 			}
+			if allowedApiClient != nil {
+				authorized := ClientAuthorized(allowedApiClient, pj)
+				if !authorized {
+					logrus.Error("client is not authorized to modify the given job")
+					continue
+				}
+			}
+			pj.Status.State = prowcrd.ProwJobState(strings.ToLower(request.GetJobStatusChange().GetDesired().String()))
+			updatedPj, err := gw.ProwJobClient.Update(context, &pj, metav1.UpdateOptions{})
+			if err != nil {
+				logrus.WithError(err).Errorf("failed to update ProwJob status")
+				continue
+			}
+			logrus.WithField("name", pj.Name).Infof("ProwJob status updated to: %s", updatedPj.Status.State)
 		}
+		cancel()
```

Comment:
Do we need this? Is it possible to just use `_` instead of creating the `cancel` var?

Reply:
yes, unfortunately, otherwise the linter job is failing